client/consensus/babe/rpc/src/lib.rs

```diff
@@ -126,22 +131,21 @@ impl<B, C, SC> BabeApi for BabeRpcHandler<B, C, SC>
 
 			let mut claims: HashMap<AuthorityId, EpochAuthorship> = HashMap::new();
 
-			let key_pairs = {
-				let keystore = keystore.read();
-				epoch.authorities.iter()
-					.enumerate()
-					.flat_map(|(i, a)| {
-						keystore
-							.key_pair::<sp_consensus_babe::AuthorityPair>(&a.0)
-							.ok()
-							.map(|kp| (kp, i))
-					})
-					.collect::<Vec<_>>()
-			};
+			let ks = keystore.read();
```

Comment:
deadlock ahead: this `ks` read-lock is held until the end of the function, and `keystore` is passed into `claim_slot_using_keys` where it may attempt to get a read lock again - this can lead to a deadlock.

I recommend moving this statement to an inner scope as it was before to scope the lifetime of the read lock here.

Reply:
@DemiMarie-parity had some performance concerns what the `read` was hapening inside the iteration. But i guess we can just have read guards created and dropped inside `filter_map` for now to prevent the deadlocks.

Reply:
Actually, an alternative is to call `drop(ks)` after we don't need this anymore.